### 【目录】

* [1. 变量及数据类型](#1. 变量及数据类型)
	* [1.1 数据类型概述](#1. 数据类型概述)
	* [1.2 认识变量](#1.2 认识变量) 
	* [1.3 运算符](#1.3 运算符)
	* [1.4 注释](#1.4 注释)
* [2. 序列及通用操作](#2. 序列及通用操作)
	* [2.1 序列通用操作](#2.1 序列通用操作)	
	* [2.2 列表list常用操作](#2.2 列表list常用操作)
	* [2.3 文本序列str常用操作](#2.3 文本序列str常用操作)
* [3. 字典映射](#3. 字典映射)
	* [3.1 字典dict基本概念](#3.1 字典dict基本概念)
	* [3.2 字典常用操作](#3.2 字典常用操作)
	* [3.3 字典的元素访问及遍历](# 3.3 字典的元素访问及遍历)
* [4. 条件及循环语句](#4. 条件及循环语句)
	* [4.1 条件判断：if语句](#4.1 条件判断：if语句)
	* [4.2 循环语句：for循环](#4.2 循环语句：for循环)
	* [4.3 循环语句：while循环](#4.3 循环语句：while循环)
	* [4.4 循环控制语句](#4.4 循环控制语句)
* [5. 函数](#5. 函数)
	* [5.1 自定义函数](#5.1 自定义函数)
	* [5.2 局部变量及全局变量](#5.2 局部变量及全局变量)
	* [5.3 匿名函数Lambda](#5.3 匿名函数Lambda)
* [6. 模块与包](#6. 模块与包)
	* [6.1 模块创建及import指令运用](#6.1 模块创建及import指令运用)
* [7. 数据读写](#7. 数据读写)
	* [7.1 文件对象说明及基本操作](#7.1 文件对象说明及基本操作)
	* [7.2 系统模块下的路径操作](#7.2 系统模块下的路径操作)
	* [7.3 文件的读取与写入](#7.3 文件的读取与写入)
	* [7.4 pickle模块的运用](#7.4 pickle模块的运用)

PS：Fork本项目点击运行后，在左面侧边栏可以看到目录，点击相应章节即可跳转。

# 1. 变量及数据类型
## 1.1 数据类型概述

Python标准数据类型：
* Numbers数字
* String字符串
* List列表
* Tuple元组
* Dict字典

In [1]:
# Numbers数字分为：int整型，long长整型，float浮点型，complex复数

x1 = 10
x2 = 10.0
print(type(x1),type(x2))

# print()函数，用于输出/打印内容
# type()函数，用于查看数据类型

<class 'int'> <class 'float'>


In [2]:
# String字符串由数字、字母、下划线组成的一串字符，用于表示文本的数据类型

x3 = "hello world!"
print(x3,type(x3))
print('hehe')
print("hahaha")
print('''what is this?
it`s a cat.''')
# 一定用英文标点符号，单引号('') 双引号("")一样，三引号(''''''或"""""")可以表示多行字符串

hello world! <class 'str'>
hehe
hahaha
what is this?
it`s a cat.


In [3]:
# bool布尔型：True，False，用于做判断

x4 = True
print(x4,type(x4))

print(True == 1)
print(False == 0)
print(True * 10)
# True实际值是1，False实际值是0

True <class 'bool'>
True
True
10


In [4]:
# List列表，支持字符，数字，字符串以包含列表（即嵌套），用[]标识，有序对象

x5 = [1,'a',2.4,502.2,[1,2,3]]
print(x5,type(x5))

[1, 'a', 2.4, 502.2, [1, 2, 3]] <class 'list'>


In [5]:
# Tuple元祖，用（）标识，不能二次赋值，可以理解成不可变的列表（只读列表），有序对象

x6 = (1,2,3,'hello')
print(x6,type(x6))

(1, 2, 3, 'hello') <class 'tuple'>


In [6]:
# Dict字典，用{}标识，由索引(key)和它对应的值value组成，无序对象

x7 = {"name":"fatbird", "city":"shanghai", "tel":10001000}
print(x7,type(x7))

{'name': 'fatbird', 'city': 'shanghai', 'tel': 10001000} <class 'dict'>


In [7]:
# 数据类型转换方法  **这里暂时只针对int，float，str三者，list/tuple/dict后面课程讲

var1 = 10.8
var2 = int(var1)
print(var1,type(var1))
print(var2,type(var2))
print('----------')
# int(x)函数：将x转为整数，如果是四舍五入则是round(x)函数

var1 = 10
var2 = float(var1)
print(var1,type(var1))
print(var2,type(var2))
print('----------')
# float(x)函数：将x转为浮点数

var1 = 10.0
var2 = str(var1)
var3 = float(var2)
print(var1,type(var1))
print(var2,type(var2))
print(var3,type(var3))
print('----------')
# str(x)函数：将x转为字符串

10.8 <class 'float'>
10 <class 'int'>
----------
10 <class 'int'>
10.0 <class 'float'>
----------
10.0 <class 'float'>
10.0 <class 'str'>
10.0 <class 'float'>
----------


## 1.2 认识变量

什么是变量？ —— 方程式“=”左边的，就是变量

1、给变量赋值即创建该变量，并且同时赋予该变量
① 变量名
② 变量类型
等号（=）用来给变量赋值

2、变量命名规则：
① 变量名第一个字符必须是字母（大小写均可）或者下划线（“_”），不能数字开头；
② 变量名不能和常用功能性名字重合，例如print，if，for
③ 不要有空格

In [8]:
# 变量赋值

counter = 100 # 赋值整型变量
miles = 1000.0 # 浮点型
name = "John" # 字符串
 
print(counter)
print(miles)
print(name)

100
1000.0
John


In [9]:
# 多变量赋值

a = b = c = 1
d , e , f = 1 , 2 , 'hello'
print(a,b,c,d,e,f)

1 1 1 1 2 hello


In [10]:
# 动态变量 → 变量的值是可更改的，或者说，变量可以随时指向任意值

var1 = 10
print(var1)
var1 = 20
print(var1)

10
20


## 1.3 运算符

什么是运算符？

1 +2 = 3 的例子中，1 和 2 被称为操作数，"+" 称为运算符

In [11]:
# 算数运算符

a, b, c = 21, 10, 0
 
c = a + b  # 加法
print ("a加b 的值为：", c)
 
c = a - b  # 减法
print( "a减b 的值为：", c )
 
c = a * b  # 乘法
print( "a乘b 的值为：", c )
 
c = a / b  # 除法
print( "a除b 的值为：", c )
 
c = a % b  # 取模，返回除法的余数
print( "a取b的模为：", c)
 
# 修改变量 a 、b 、c
a, b, = 2, 3
c = a**b   # 幂
print( "2的3次方的值为：", c)
 
a, b = 10, 4
c = a//b  # 取整除数 
print( "10除4的整除数为：", c)

a加b 的值为： 31
a减b 的值为： 11
a乘b 的值为： 210
a除b 的值为： 2.1
a取b的模为： 1
2的3次方的值为： 8
10除4的整除数为： 2


In [12]:
# 比较运算符

a, b = 21, 10
print(a == b)  # 等于
print(a != b)  # 不等于
print(a > b)  # 大于
print(a < b)  # 小于
print(a >= 21)  # 大于等于
print(a <= 2)  # 小于等于

False
True
True
False
True
False


In [13]:
# 逻辑运算符

print(True and False)  # and：且
print(True or False)  # or：或
print(not True)  # not：非
print('-----')

# 布尔型的本质：True的值等于1，False的值等于0
print(True>0)
print('-----')

print(bool(0),bool(3),bool(-1),bool([1,2,3]),bool(0))
# bool()函数：将值转换为布尔型，其中只有以下情况bool()返回False：0, 0.0, None，'', [] , {}

False
True
False
-----
True
-----
False True True True False


In [14]:
# 成员运算符

lst = [1,2,3,4,5,6]
a,b = 1,10
print(a in lst)  # a 在 lst 序列中 , 如果 x 在 y 序列中返回 True。
print(b not in lst)  # b 不在 lst 序列中 , 如果 x 不在 y 序列中返回 True。

True
True


## 1.4  注释#

* 注释是用来说明代码
* 程序运行的时候，Python解释器会直接忽略掉注释 → 有没有注释不影响程序的执行结果，但是影响到别人能不能看懂你的代码。
* 但“#”不只代表注释,还代表某些文件的特殊格式，写在脚本开头
* 注意：注释只在脚本中起作用

In [15]:
print('hello world!')
# print('I love python')

hello world!


In [16]:
a = [1,2,3,4]
#a = {'name':'fatbird', 'city':'shanghai', 'tel':10001000}
for i in a:
    print(i)

# 注释也可以用于标记暂不运行语句

1
2
3
4


# 2. 序列及通用操作
## 2.1 序列通用操作

序列分类：
* 可变序列list
* 不可变序列tuple、str

In [17]:
# 判断值是否属于序列

lst = [1,2,3,4,5,6]
a,b = 1,10
print(a in lst)  # a 在 lst 序列中 , 如果 x 在 y 序列中返回 True。
print(b not in lst)  # b 不在 lst 序列中 , 如果 x 不在 y 序列中返回 True。

True
True


In [18]:
# 序列链接与重复

lst1 = [1,2,3]
lst2 = ['a','b','c']
print(lst1+lst2)  # "+"：序列的链接
print(lst1*3,lst2*2)  # "*"：序列重复

[1, 2, 3, 'a', 'b', 'c']
[1, 2, 3, 1, 2, 3, 1, 2, 3] ['a', 'b', 'c', 'a', 'b', 'c']


In [19]:
# 下标索引

lst = [1,2,3,4,5,6,7,8,9,0]
print(lst[0],lst[2],lst[9])  # 索引从0开始，eg lst[2]中，下标索引是2，指向lst的第3个值
print(lst[-1])  # 索引-1代表最后一个值
#print(lst[10])  # 这里一共有10个值，所以最大索引是9

1 3 0
0


In [20]:
# 切片

lst = [1,2,3,4,5,6,7,8,9,0]
print(lst[2:5])  # 切片可以理解成列表的值区间，且是一个左闭右开区间，这里lst[2:5]代表的区间是：索引2的值 - 索引4的值
print(lst[:5])  # 左边无索引，代表从索引0开始
print(lst[4:])  # 右边无索引，代表以最后一个值结束
print(lst[5:-1])  # 索引5的值 - 倒数第二个值 （切片是右闭区间，注意了）

[3, 4, 5]
[1, 2, 3, 4, 5]
[5, 6, 7, 8, 9, 0]
[6, 7, 8, 9]


In [21]:
# 步长

lst = [1,2,3,4,5,6,7,8,9,0]
print(lst[0:5:2])  # List[i:j:n]代表：索引i - 索引j，以n为步长
print(lst[::2])  # 按照2为步长，从第一个值开始截取lst数据
print(lst[1::2])  # 按照2为步长，从第二个值开始截取lst数据

[1, 3, 5]
[1, 3, 5, 7, 9]
[2, 4, 6, 8, 0]


In [22]:
# 序列的基本内置全局函数

lst = [1,2,3,4,5,6,7,8,9,0]
print(len(lst))  # 列表元素个数
print(max(lst),min(lst),sum(lst))  # 返回列表的最大值、最小值、求和，这三个函数都只针对数字的list
print(lst.index(3))  # .index(obj)方法：从列表中找出某个值第一个匹配项的索引位置

lst = [1,1,2,3,3,4,4,4,4,5,6]
print(lst.count(4))  # .count(obj)方法：计算值的出现次数

10
9 0 45
2
4


小作业

① x=[1,2,3,1,[1],[1,1]]，请问x.count(1)为多少？

② x=[1,2,5,1,“1”，“1”]，请问x.count(1)为多少？

③ y=[45,67,89,45,34,22,1],请问y.index(45)等于多少？

④ 创建一个列表，用type函数判断list[1]和list[2:5]的区别

## 2.3 列表list常用操作

* 序列是Python中最基本的数据结构，它支持字符，数字，字符串甚至可以包含列表（即嵌套）
* 列表用 [ ] 标识，是 python 最通用的复合数据类型。
* 列表中值的切片也可以用到变量 [头下标:尾下标] ，就可以截取相应的列表，从左到右索引默认 0 开始，从右到左索引默认 -1 开始，下标可以为空表示取到头或尾。
* 加号 + 是列表连接运算符，星号 * 是重复操作



In [23]:
# 列表的特征

lst1 = [1,2,3,'a','b','c',[1,2,3]]
# 可包含任意类型的对象：数值、字符串、列表等

lst2 = [143,56,894,67,43]
print(lst2[0],lst2[4],lst2[-1])
# 通过下标索引访问序列中的值 → 序列的顺序是不能改变的,通过索引来定位列表的元素

lst3 = [1,2,3]
lst3 = lst3 * 3
print(lst3)
# 可随意变换列表长度，相当于随意指定新的列表

lst4 = [1,2,['a','b']]
# 可嵌套

lst5 = [100,101,102]
lst5[0] = 10
print(lst5)
# 可原位改变

143 43 43
[1, 2, 3, 1, 2, 3, 1, 2, 3]
[10, 101, 102]


小作业

① [1,2,3]和[“1”, “2”, “3”]什么区别

② 创建一个列表，包括3个人的数学成绩100，89,78

③ 创建一个列表，包括5个中国城市名称

④ 现有一个列表name = ['小明', '小红'] ,增加“团支书”的名字上去

⑤ 现有一个列表m=[3.1,5.5,4,3,77,8]，用print分别输出第2、4个元素

In [24]:
# 列表与生成器

print(range(5),type(range(5)))  
# range()是生成器，指向了一个范围
# range(5)代表指向了0,1,2,3,4这几个值
# range(2,5)代表指向了2,3,4这几个值，注意这里不是使用:
# range(0,10,2)代表指向了0,2,4,6,8这几个值，最后的2代表步长

lst = list(range(5))
print(lst)
# 通过list()函数生成列表

range(0, 5) <class 'range'>
[0, 1, 2, 3, 4]


小作业

① range()生成的是不是都是整数?

② 创建一个range，用type()函数查看其类型

③ 哪种方式书写正确：x=range(5), x=range(4:7), x=range(4,8)

④ 创建一个range，用print输出其值，看一下输出结果

⑤ 用list()函数，将一个range范围变为列表

⑥ m=range(10),请问m[2],m[-1],m[5]是多少

In [25]:
# 可变列表list常用操作 - 添加元素

lst = list(range(10))
lst.append('hello')
print(lst)
# list.append() 添加元素

lst.append(['a','b','c'])
print(lst)
lst.extend(['a','b','c'])
print(lst)
# 添加多个元素用.extend()方法：用新列表扩展原来的列表，注意和append()方法的区别

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'hello']
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'hello', ['a', 'b', 'c']]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'hello', ['a', 'b', 'c'], 'a', 'b', 'c']


In [26]:
# 可变列表list常用操作 - 删除元素

lst = ['Jack','Jack','Tom','Tracy','Alex','White']
lst.remove('Jack')
print(lst)
# .remove方法：移除列表中某个值的第一个匹配项

del lst[3:5]
print(lst)
# del语句：删除list的相应索引值

lst.clear()
print(lst)
# 移除所有值

['Jack', 'Tom', 'Tracy', 'Alex', 'White']
['Jack', 'Tom', 'Tracy']
[]


In [27]:
# 可变列表list常用操作 - 插入元素

lst = list(range(10))
print(lst)
lst.insert(3,'a')
print(lst)
# x.insert(i,m)方法：在索引i处插入m，这里索引3代表第四个值

lst.insert(5,[1,2,3])
print(lst)
# 插入一个列表元素

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 'a', 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 'a', 3, [1, 2, 3], 4, 5, 6, 7, 8, 9]


In [28]:
# 可变列表list常用操作 - 复制

lst = list(range(10))
lst1 = lst
lst1[2] = 'hello'
print(lst,lst1)
# lst，lst1指向同一个列表

lst = list(range(10))
lst2 = lst.copy()
lst2[2] = 'hello'
print(lst,lst2)
# list.copy()方法：复制一个新的列表，lst，lst2指向两个列表（虽然两个列表值相同）

[0, 1, 'hello', 3, 4, 5, 6, 7, 8, 9] [0, 1, 'hello', 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9] [0, 1, 'hello', 3, 4, 5, 6, 7, 8, 9]


In [29]:
# 可变列表list常用操作 - 排序

lst1 = [12,45,32,98,45,66,3]
lst2 = ["asd", "cxv", "ghv"]
lst1.sort()
lst2.sort()
print(lst1,lst2)
# 默认升序排序/字母顺序

lst1.sort(reverse=True)
lst2.sort(reverse=True)
print(lst1,lst2)
# 参数reverse：反向排序，针对数字

lst3 = sorted(lst1)
lst3.append('hello')
print(lst1,lst3)
# 函数sorted():排序并复制

[3, 12, 32, 45, 45, 66, 98] ['asd', 'cxv', 'ghv']
[98, 66, 45, 45, 32, 12, 3] ['ghv', 'cxv', 'asd']
[98, 66, 45, 45, 32, 12, 3] [3, 12, 32, 45, 45, 66, 98, 'hello']


In [30]:
# 讲一讲元祖tuple ：不可变的“列表”

tup1 = ('physics', 'chemistry', 1997, 2000);
tup2 = (1, 2, 3, 4, 5 );
tup3 = "a", "b", "c", "d"  # 可以不加括号
tup4 = (50,)  # 元组中只包含一个元素时，需要在元素后面添加逗号

print(tup1[2],tup3[:2])  # 索引、切片和list一样

del tup3 # 元祖不能单独删除内部元素（不可变性），但可以del语句来删除整个元祖
#print(tup3)

print(len(tup1))
print(tup1 + tup2)
print(tup4 * 3)
print(max(tup2),min(tup2))
# 序列通用函数

lst = list(range(10))
tup5 = tuple(lst)
print(tup5)
# tuple()函数：将列表转换为元组

1997 ('a', 'b')
4
('physics', 'chemistry', 1997, 2000, 1, 2, 3, 4, 5)
(50, 50, 50)
5 1
(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)


小作业

① 分别用append、extend方法，给一个序列增加一个元素1

② x.append([1,2,3])和x.extend([1,2,3])的区别

③ x.append('abc')和x.extend('abc')的区别

④ 创建某一列表，在列表中间插入一个值

⑤ 创建一个随机数值列表，按照降序排列

⑥ sort()和sorted()什么区别

⑦ s=[32,45,63,12,45,89,77], s.remove(45),  del s[2], s.pop(4)分别什么区别

⑧ m=[4,5,6,73,4,4,4,4,2]，删除2，删除所有的4

## 2.3 文本序列str常用操作

* 字符串是 Python 中最常用的数据类型。我们可以使用引号('或")来创建字符串
* 字符串也是序列：文本序列

In [31]:
# 字符串引号

str1 = "abc"
str2 = 'abc'
str3 = 'my name is "fatbird"'
# 双引号单引号无区别，但文本中有引号的时候要相互交替使用

str4 = '''hello! how are you?
          I`m fine, thank you.'''
# 需要多行字符串时候用三引号 ''' ''',""" """

In [32]:
# 转义字符：\

print('\'', '\"')  # \',\" ：分别输出单引号，双引号
print('hello\nhow do you do?')  # \n：空行
print('\\')  # 输出\，所以文件路径用“/”或者“\\”

' "
hello
how do you do?
\


In [33]:
# 序列通用功能

print('a' in 'abc')  # in/not in ：判断是否存在
print('我很帅' + "没错")  # 文本连接
print('handsome '*4)  # 文本复制

st = 'abcdefg'
print(st[2],st[-1])
print(st[:2])
print(st[::2])
# 索引、切片、步长

print(st.index('g'))  # .index()方法
print('st长度为：',len(st))  # 计算字符串长度（思考这里能否把逗号换为"+"）

True
我很帅没错
handsome handsome handsome handsome 
c g
ab
aceg
6
st长度为： 7


In [34]:
# 字符串常用功能

st = "i`m handsome!"
st2 = st.replace('handsome','ugly')
print(st)
print(st2)
st = 'hahaha'
st2 = st.replace('ha','he',2)
print(st2)
# str.replace(old,new,count)：修改字符串，count：更换几个

st = "poi01,116.446238,39.940166"
lst = st.split(',')
print(lst)
# str.split(obj)：拆分字符串，生成列表

lst = ['poi01', '116.446238', '39.940166']
m = '-'
st = m.join(lst)
print(st)
# str.join()：连接字符串，对象为列表

st = 'abcdefg'
print(st.startswith('a'), st.endswith('f'))
# str.startswith(“str”)  判断是否以“str”开头；str.endswith (“str”)  判断是否以“str”结尾

st = 'aBDEAjc kLM'
print(st.upper())  # 全部大写
print(st.lower())  # 全部小写
print(st.swapcase())  # 大小写互换
print(st.capitalize())  # 首字母大写

st = '1234567'
print(st.isnumeric())  # 如果 string 只包含数字则返回 True,否则返回 False.

st = 'DNVAK'
print(st.isalpha())  # 如果 string 至少有一个字符并且所有字符都是字母则返回 True,否则返回 False

st = 'avd   '
print(st.rstrip())  # 删除字符末尾的空格

i`m handsome!
i`m ugly!
heheha
['poi01', '116.446238', '39.940166']
poi01-116.446238-39.940166
True False
ABDEAJC KLM
abdeajc klm
AbdeaJC Klm
Abdeajc klm
True
True
avd


In [35]:
# 格式化字符：在字符串中插入变量

name = "FATBIRD"
truth = "%s is 好人" %name
print(truth)
# %s 表示我们这里要插入一个变量（字符串），%name代表这里我们插入name → 所以，两个%解决问题

x = 4   
y = "hehe"  
z = 4.2
print("this is %i" %x)
print("this is %s" %y)
print("this is %f" %z)
# %s的s有什么含义：%s %z %f都是格式化字符串，s/i/f分别代表插入字符串的源数据类型：字符型/整型/浮点型

FATBIRD is 好人
this is 4
this is hehe
this is 4.200000


In [36]:
# 格式化字符：数字格式化的那些坑

m = 3.1415926 
print("pi is %f"  %m)
print("pi is %.2f" %m)
# 我只想输出2位小数:%.2f，此处是四舍五入！

m = 10.6
print("pi is %i"  %m)
print("pi is %.0f" %m)
# 区别：%i 不四舍五入，直接切掉小数部分

m = 100 
print("have fun %+i"  %m)
print("have fun %.2f"  % -0.01)
# 显示正号，负号根据数字直接显示

m = 100 
print("have fun % i"  %m)
print("have fun % +i"  %m)
print("have fun % .2f"  %-0.01)
# 加空格，空格和正好只能显示一个

m = 123.123123123 
print("have fun %.2e"  %m)
print("have fun %.4E"  %m)
# 科学计数法 %e  %E

m1 = 123.123123123
m2 = 1.2
print("have fun %g"  %m1)
print("have fun %g"  %m2)
# 小数位数少的时候自动识别用浮点数，数据复杂的时候自动识别用科学计数法

pi is 3.141593
pi is 3.14
pi is 10
pi is 11
have fun +100
have fun -0.01
have fun  100
have fun +100
have fun -0.01
have fun 1.23e+02
have fun 1.2312E+02
have fun 123.123
have fun 1.2


In [37]:
# 更强大的格式化方法.format

print("User ID: {0}".format("root"))
print("{} 呵呵 {}".format("a","b"))
# {} 这里代表占位符，其中可以有数字也可以没有

print("{}{}{}".format('a','b','c'),'\n',
     "{0}{1}{2}{0}".format('a','b','c'),'\n')
#print("{}{}{}{}".format('a','b','c'))
# {}和{0}的区别:都是占位符，后者有了明确指定

print("我的工作是{work}".format(work = '设计'))
# 也可以用变量来指示

x="abc{}"
a = x.format("def")
print(a,x)
# .format()生成新的值吗？？

print("{:f}".format(4.123),'\n',
 "{:.2f}".format(4.123),'\n',
 "{:e}".format(4.123),'\n',
 "{:.0f}".format(99.9),'\n',
 "{:%}".format(4.123),'\n',
 "{:d}".format(10))

User ID: root
a 呵呵 b
abc 
 abca 

我的工作是设计
abcdef abc{}
4.123000 
 4.12 
 4.123000e+00 
 100 
 412.300000% 
 10


小作业

① 三引号创建一个有提行的文本，print()函数输出

② 申明一个字符串变量，值为一个文件路径，注意操作习惯

③ 以下输出分别是多少：33+"22"; 33+int("22"); "22" + str(55); 

④ m=“a,b,c”,m.split()会输出什么结果？（不加","）

⑤ 再回想一下：.split和.join的输出分别是什么类型的数据

⑥ 如果我想打%怎么办？

⑦ 这样书写正确吗，为什么？ print("abc%s") % "nn"

⑧ 这样书写正确吗，为什么？ print(245%f % 123)

⑨ "我的工作是…我喜欢…" 把…替换成正确的内容 

# 3. 字典映射
## 3.1 字典dict基本概念

* 字典是另一种可变容器模型，且可存储任意类型对象。
* 字典的每个键值(key=>value)对用冒号(:)分割，每个对之间用逗号(,)分割，整个字典包括在花括号({})中 
* 键必须是唯一的，但值则不必。

In [38]:
# 什么是映射？什么是字典dict？

name = ['大头儿子','小头爸爸','隔壁老王']
age = [6,35,37]
# 首先虽然序列是有顺序的，但还是不能直接把名字和年龄关联起来，这里只是顺序对上，且并没有一对一的关系

dic = {'大头儿子':6, '小头爸爸': 35, '隔壁老王':37}
print(dic,type(dic))
# 用{}来表示，按照key：value来表示字典中的元素，其中key是键，value是值，key-value也就是键值对
# 这里dic就是一个字典类型的变量

dic = [{'name':'Tom', 'age':18},{'name':'Jack', 'age':19},{'name':'Alex', "age":17}]
print(dic)
# 一个以字典为元素的列表，有没有觉得键值对很像数据中的字段和字段值？

{'大头儿子': 6, '小头爸爸': 35, '隔壁老王': 37} <class 'dict'>
[{'name': 'Tom', 'age': 18}, {'name': 'Jack', 'age': 19}, {'name': 'Alex', 'age': 17}]


In [39]:
# 字典的特点

dic = {'a':1,'b':2,'c':3}
# 键值对，一定共同出现，不能只有一个

dic = {'a':1,'a':2}
print(dic)
# 不允许同一个键出现两次：创建时如果同一个键被赋值两次，后一个值会被记住

dic = {'var1':1, 'var2':'hello', 'var3':[1,2,3], 'var4':{'a':1,'b':2}}
# key必须是字符串，value则可以是任意对象：数值，字符串，序列，字典

print(dic)
# 字典是一个无序集合,序列由于没有key来做对应，所以以顺序来对应值
# 序列有顺序，字典没有！

dic['var5'] = '添加一个值'  # 任意添加元素
dic['var4'] = '任意修改其中一个值'  # 任意修改元素
del dic['var1']  # 任意删除元素
print(dic)

dic.clear()  # 清空词典所有条目
print(dic)

del dic  # 删除词典
#print(dic)  

{'a': 2}
{'var1': 1, 'var2': 'hello', 'var3': [1, 2, 3], 'var4': {'a': 1, 'b': 2}}
{'var2': 'hello', 'var3': [1, 2, 3], 'var4': '任意修改其中一个值', 'var5': '添加一个值'}
{}


In [40]:
# 申明一个字典

dic = {'a':1,'b':2}
# 最简单直接的方式，手动书写申明一个字典变量，注意是：不是=

dic = dict(m = 10 ,n = 'aa', h = [1,2,3])
print(dic)
# dict()函数

lst1=[("a","fff"),("b","ggg")]
lst2=[["c",1],["d",2]]
lst3=(("e",[1,2]),("f",[3,4]))
dic1 = dict(lst1)
dic2 = dict(lst2)
dic3 = dict(lst3)
print(dic1,dic2,dic3)
# dict()由序列生成一个字典，嵌套序列，可以是list或者tuple

keys = ["a", "b", "c"]
dic1 = dict.fromkeys(keys)
dic2 = dict.fromkeys(keys, 'hello')
print(dic1,dic2)
# 只有key，直接生成一个字典

{'m': 10, 'n': 'aa', 'h': [1, 2, 3]}
{'a': 'fff', 'b': 'ggg'} {'c': 1, 'd': 2} {'e': [1, 2], 'f': [3, 4]}
{'a': None, 'b': None, 'c': None} {'a': 'hello', 'b': 'hello', 'c': 'hello'}


小作业

① 用dict函数的时候，和直接申明一个字典，操作上什么区别？

② 用dict()函数从序列创建字典的时候，对序列有要求吗？必须是列表吗？

③ 字典是无序的，强化一下

④ 以dict()函数生成一个人的成绩单，包括语文，数学，外语，物理，化学几个元素

## 3.2 字典常用操作

* 字典内置函数&方法

In [41]:
# 字典常用操作

dic1 = {"a":1, "b":2}
dic2 = {"c":3, "d":4}
dic1.update(dic2)
print(dic1,dic2)
# dict.update()方法：更新/合并一个字典，把第二个字典合并到第一个字典，改变了第一个字典

a = {'m':1 , 'n':2 , 'p':3}
b = a
a.update({'q':4})
print(a,b)
# 和序列的原理一样，a和b指向同一个字典，所以会一起更新

a = {'m':1 , 'n':2 , 'p':3}
b = a.copy()
a.update({'q':4})
print(a,b)
# 通过.copy()方法，复制一个新的字典

print(len(a))
# 查看字典元素个数

print('m' in a)
print(1 in a)
# in / not in ：判断是否包含，这里的判断对象是key

{'a': 1, 'b': 2, 'c': 3, 'd': 4} {'c': 3, 'd': 4}
{'m': 1, 'n': 2, 'p': 3, 'q': 4} {'m': 1, 'n': 2, 'p': 3, 'q': 4}
{'m': 1, 'n': 2, 'p': 3, 'q': 4} {'m': 1, 'n': 2, 'p': 3}
4
True
False


## 3.3 字典的元素访问及遍历

* 如何读取字典？

In [42]:
# 字典的元素访问

dic = {'a':1, "b":2, "c":3}
print(dic['a'])
#print(dic[1])  # 不能用序列的索引方式
# 字典里面也就同样的意思，但字典没有顺序，以key来作为指向，所以指向的key必须存在

poi = {'name':'shop', 'city':'shanghai', 'information':{'address':'somewhere', 'num':66663333}}
print(poi['information']['address'])
# 对于嵌套字典，输出嵌套内容，通过重复指向来输出

print(poi.get('name'))
print(poi.get('type',print('nothing')))
# .get(key)方法：直接查看key的value，如果没有相应key则返回None，添加print参数可以多返回一个值

print(poi.keys(),type(poi.keys()))
print(list(poi.keys()))
# .keys()方法：输出字典所有key，注意这里的输出内容格式是视图，可以用list()得到key的列表，类似range()

print(poi.values(),type(poi.values()))
print(list(poi.values()))
# .values()方法：输出字典所有values，原理同.keys()方法

print(poi.items(),type(poi.items()))
print(list(poi.items()))
# .items()方法：输出字典所有items（元素），原理同.keys()方法

1
somewhere
shop
nothing
None
dict_keys(['name', 'city', 'information']) <class 'dict_keys'>
['name', 'city', 'information']
dict_values(['shop', 'shanghai', {'address': 'somewhere', 'num': 66663333}]) <class 'dict_values'>
['shop', 'shanghai', {'address': 'somewhere', 'num': 66663333}]
dict_items([('name', 'shop'), ('city', 'shanghai'), ('information', {'address': 'somewhere', 'num': 66663333})]) <class 'dict_items'>
[('name', 'shop'), ('city', 'shanghai'), ('information', {'address': 'somewhere', 'num': 66663333})]


In [43]:
# 字典的元素遍历

poi = {'name':'shop', 'city':'shanghai', 'information':{'address':'somewhere', 'num':66663333}}
for key in poi.keys():
    print(key)
print('-------')

for value in poi.values():
    print(value)
print('-------')

for (k,v) in poi.items():
    print('key为 %s, value为 %s' %(k,v))
print('-------')  
# for函数遍历

name
city
information
-------
shop
shanghai
{'address': 'somewhere', 'num': 66663333}
-------
key为 name, value为 shop
key为 city, value为 shanghai
key为 information, value为 {'address': 'somewhere', 'num': 66663333}
-------


小作业

① dict.keys()生成的是不是列表？

② 如何判断一个value是否存在于字典中？

# 4. 条件及循环语句
## 4.1 条件判断：if语句

Python条件语句是通过一条或多条语句的执行结果（True或者False）来决定执行的代码块。

if 判断条件：
    执行语句……
else：
    执行语句……
    
** 开始有缩进的概念

In [44]:
# 基本判断语句

age = 12
if age < 18:
    print('18岁以下不宜观看')
# if语句后面必须有 ： 
# 自动缩进
# if语句写完后，要退回原有缩进继续写
# Python代码的缩进规则：具有相同缩进的代码被视为代码块

18岁以下不宜观看


In [45]:
# 输入函数 input()

score = input('请输入成绩：')
print('该学生成绩为：' + score)
print(type(score))
# 注意：input()返回结果都为字符串，如果需要变为数字则用到int()/float()

请输入成绩：: 100

该学生成绩为：100
<class 'str'>


In [46]:
# 两种条件判断:if-else

flag = False
name = 'luren'
if name == 'python':          # 判断变量否为'python'
    flag = True               # 条件成立时设置标志为真
    print( 'welcome boss')    # 并输出欢迎信息
else:
    print(name)               # 条件不成立时输出变量名称

luren


In [47]:
# 多种条件判断:if-elif-...-else

num = 2    
if num == 3:            # 判断num的值
    print('boss')       
elif num == 2:
    print('user')
elif num == 1:
    print('worker')
elif num < 0:           # 值小于零时输出
    print('error')
else:
    print('roadman')    # 条件均不成立时输出

user


In [48]:
# 单语句多条件判断：or and

num = 5
if num >= 0 and num <= 10:    
    print( 'hello')
# 判断值是否在0~10之间
# 输出结果: hello
 
num = 10
if num < 0 or num > 10:    
    print( 'hello')
else:
    print( 'undefine')
# 判断值是否在小于0或大于10
# 输出结果: undefine
 
num = 8
if (num >= 0 and num <= 5) or (num >= 10 and num <= 15):    
    print( 'hello')
else:
    print( 'undefine')
# 判断值是否在0~5或者10~15之间
# 输出结果: undefine

hello
undefine
undefine


小作业

① 编写一个简单的判断语句代码：输入某个成绩，如果成绩分数大于或等于60分，则返回及格，小于60分，则返回不及格

② 编写猜数字小游戏的代码：输入一个数字，分别针对猜对数字、猜错数字、输入错误给予判断

## 4.2 循环语句：for循环

for循环可以遍历任何序列的项目，如一个列表或者一个字符串。

迭代

In [49]:
# 想输出"hello world"5次怎么办？
for i in range(5):
    print('hello world!')

hello world!
hello world!
hello world!
hello world!
hello world!


In [50]:
# 通过for遍历序列、映射

lst = list(range(10))
for i in lst[::2]:
    print(i)
print('-----')
# 遍历list

age = {'Tom':18, 'Jack':19, 'Alex':17, 'Mary':20}
for name in age:
    print(name + '年龄为：%s岁' % age[name])
# 遍历字典

0
2
4
6
8
-----
Tom年龄为：18岁
Jack年龄为：19岁
Alex年龄为：17岁
Mary年龄为：20岁


In [51]:
# 嵌套循环

for i in range(3):
    for j in range(2):
        print(i,j)
# 循环套循环，注意：尽量不要多于3个嵌套

0 0
0 1
1 0
1 1
2 0
2 1


小作业

① 生成一个数值列表，用for循环打印出所有元素

② 用for循环遍历一个字符串，打印出各个字母

③ 生成一个字典，分别打印出key和value

④ 用input输入一个循环次数n，打印hello world n遍

⑤ 码一个等差数列，四个变量：首项a，项数n，公差d，求和s，这几个参数都可通过input()输入

⑥ 两组列表["a", "b", "c"],[1,2,3]，用for循环把它们组成一个字典，一一对应

## 4.3 循环语句：while循环

执行语句可以是单个语句或语句块

判断条件可以是任何表达式，任何非零、或非空（null）的值均为true。

当判断条件假false时，循环结束。

In [52]:
# 基本运行逻辑

count = 0
while count < 9:
    print( 'The count is:', count)
    count = count + 1
print( "Good bye!")
# 这里count<9是一个判断语句，当判断为True时，则继续运行

The count is: 0
The count is: 1
The count is: 2
The count is: 3
The count is: 4
The count is: 5
The count is: 6
The count is: 7
The count is: 8
Good bye!


In [53]:
# 关于无限循环：如果条件判断语句永远为 true，循环将会无限的执行下去

# var = 1
# while var == 1 :  
#     num = input("Enter a number  :")
#     print( "You entered: ", num)
# print( "Good bye!")

# 该条件永远为true，循环将无限执行下去
# 一定要避免无限循环！！

In [54]:
# while-else语句

count = 0
while count < 5:
    print(count, " is  less than 5")
    count = count + 1
else:
    print(count, " is not less than 5")
# 逻辑和if-else一样

0  is  less than 5
1  is  less than 5
2  is  less than 5
3  is  less than 5
4  is  less than 5
5  is not less than 5


## 4.4 循环控制语句

* break：在语句块执行过程中终止循环，并且跳出整个循环
 
* continue：在语句块执行过程中跳出该次循环，执行下一次循环

* pass：pass是空语句，是为了保持程序结构的完整性

In [55]:
# break语句

s = 0
n = 1
while n > 0:
    s = s + n
    n = n + 1
    if n == 20:
        break
print(s)
# break语句用来终止循环语句，即便循环条件没有False条件或者序列还没被完全递归完，也会停止执行循环语句。

s = 0
for i in range(10):
    for j in range(5):
        s = s + (i*j)
        print('第%i次计算' %(i+j))
    if s > 20:
        break
print('结果为%i' % s)
# 如果使用嵌套循环，break语句将停止执行最深层的循环，并开始执行下一行代码。

190
第0次计算
第1次计算
第2次计算
第3次计算
第4次计算
第1次计算
第2次计算
第3次计算
第4次计算
第5次计算
第2次计算
第3次计算
第4次计算
第5次计算
第6次计算
结果为30


In [56]:
# continue语句

s = 0
for i in range(50):
    if i%2 == 0:
        s += i
    else:
        continue
    print('第%i次计算'%(i/2))
print('结果为%i' % s)
# continue 语句用来告诉Python跳过当前循环的剩余语句，然后继续进行下一轮循环。

第0次计算
第1次计算
第2次计算
第3次计算
第4次计算
第5次计算
第6次计算
第7次计算
第8次计算
第9次计算
第10次计算
第11次计算
第12次计算
第13次计算
第14次计算
第15次计算
第16次计算
第17次计算
第18次计算
第19次计算
第20次计算
第21次计算
第22次计算
第23次计算
第24次计算
结果为600


In [57]:
# pass语句

for letter in 'Python':
    if letter == 'h':
        pass
        print( '当前字母 : h，但是我pass了')
    print( '当前字母 :', letter)
print( "Good bye!")
# pass是空语句，是为了保持程序结构的完整性。（不中断也不跳过）

当前字母 : P
当前字母 : y
当前字母 : t
当前字母 : h，但是我pass了
当前字母 : h
当前字母 : o
当前字母 : n
Good bye!


# 5.函数
## 5.1 自定义函数

* 如何创建函数？def语句

In [58]:
# 定义函数

def f(x):
    if x < 5:
        print('输入值小于5')
    else:
        print('输入值大于等于5')  # 定义函数，其中x是参数（局部变量）
f(10)  # 运行函数

输入值大于等于5


In [59]:
# 关于retuen

def f1(x):
    y = 2**x  # 没有return
    
def f2(x):
    y = 2**x
    return y # 含有return

print(f1(2),f2(2))
# return语句退出函数，并返回一个表达式。不带参数值的return语句返回None

None 4


In [60]:
# 默认参数

def f(x,n = 2):
    return(x**n)
print(f(10))
print(f(10,3))
# n = 2,这里n的默认值为2，如果不输入则以默认值为主

100
1000


In [61]:
# 可变参数

def f(*x):
    print(x)
    return x
f(1)
f('a','b')
f(1,2,3,[44,33])
print(type(f('a','b')))
# 通过*来定义可变参数
# 默认会把可变参数传入一个元祖！

(1,)
('a', 'b')
(1, 2, 3, [44, 33])
('a', 'b')
<class 'tuple'>


小作业

① 函数f（x），输入一个字符串，分别print出每个字母

② f(x,y,z),函数内部算法：生成 ((x+y)*(x-y))**z

③ 编写一个求平均值的函数  f(*m)

④ 定义一个函数，用于求矩形面积、圆形面积

⑤ 定义一个函数，函数的作用是把输入的列表变成一连串字典的key，并生成字典，需要用input输入

## 5.2 局部变量及全局变量

* 定义在函数内部的变量拥有一个局部作用域，定义在函数外的拥有全局作用域。

* 局部变量只能在其被声明的函数内部访问，而全局变量可以在整个程序范围内访问。调用函数时，所有在函数内声明的变量名称都将被加入到作用域中

In [62]:
# 演示案例

def f(m):
    m = '呵呵哒'  # 函数作用：把输入变量指向“呵呵哒”
    print("函数内为局部变量：%s" % m)    

a = input('请输入一个数字：')
f(a)
print("函数外为全局变量：%s" % a)
# f（m）中，m是函数的参数，f(x)是吧x的值赋予了m，但x值自己本身不受影响，所以执行函数后，是在函数局部“变化”了x
# 什么是局部变量？ → 当函数定义内声明变量的时候，它们与函数外具有相同名称的其他变量没有任何关系!
# 即变量名称对于函数来说是“局部”的。这称为变量的作用域。所有变量的作用域是它们被定义的块，从它们的名称被定义的那点开始。

请输入一个数字：: 9

函数内为局部变量：呵呵哒
函数外为全局变量：9


In [63]:
# global语句 → 把局部变量变为全局变量

def f():
    global a
    a = '呵呵哒'  
    print("函数内转为全局变量：%s" % a)    

a = input('请输入一个数字：')
print("输入变量值：%s" % a)
f()

请输入一个数字：: 9

输入变量值：9
函数内转为全局变量：呵呵哒


## 5.3 匿名函数Lambda

* lambda 语句创建匿名函数

In [64]:
# 示例

def fun(a,b,c):
    return a+b+c
print(fun(2,3,4))
# def创建自定义函数，求和

f = lambda a,b,c:a+b+c
print(f(2,3,4))
# lambda作为匿名函数，是不需要函数名的 → 需要一个函数，又不想动脑筋去想名字，这就是匿名函数
# lambda的主体是一个表达式，而不是一个代码块。
# lambda只有一行，仅仅能在lambda表达式中封装有限的逻辑进去。

9
9


小作业

① 用lambda创建一个求元素个数的匿名函数

② 定义一个函数，可将输入的所有数字从大到小依次排序

③ 定义一个函数，随机输入一个学生的成绩后，学习成绩>=90分的同学用A表示，60-89分之间的用B表示，60分以下的用C表示，最后输出成绩评分

④ 定义一个函数，可统计出输入任意的字符中英文字母、空格、数字和其它字符的个数

# 6. 模块与包
## 6.1 模块创建及import指令运用

* Python 模块(Module)，是一个 Python 文件，以 .py 结尾，包含了 Python 对象定义和Python语句

PS：本章适用于使用本地环境的同学，K-Lab kernel中内置各类主流的机器学习、深度学习框架，集成了常见数据分析包，用户可以免除前期环境配置，直接开展分析工作。
详情请查看：https://www.kesci.com/home/workspace/help

In [65]:
# 创建一个模块，包含一个阶乘函数f1(n)、一个列表删值函数f2(lst,x),一个等差数列求和函数f3(a,d,n)

def f1(n):
    y = 1
    for i in range(1,n+1):
        y = y * i
    return y
# 创建阶乘函数f1(n)

def f2(lst,x):
    while x in lst:
        lst.remove(x)
    return lst
# 创建列表删值函数f2(lst,x)

def f3(a,d,n):
    an = a
    s = 0
    for i in range(n-1):
        an = an + d
        s = s + an
    return s
# 创建等差数列求和函数f3(a,d,n)
# 创建模块testmodel2，包括三个函数

In [66]:
# 模块路径问题

import pandas
print(pandas.__file__)
# 查看现有包所在路径，将自己创建的包存入改路径

import sys
sys.path.append('C:/Users/Hjx/Desktop/')
# 加载sys包，把新建的testmodel所在路径添加上

In [67]:
# 调用模块语句：import

import testmodel2

print(testmodel2.f1(5))
print(testmodel2.f2([2,3,4,5,5,5,6,6,4,4,4,4],4))
print(testmodel2.f3(10,2,10))
# 直接用import调用模块，.f1()调用模块函数（方法）

In [68]:
# 简化模块名：import...as...

import testmodel2 as tm2

print(tm2.f1(5))
# 简化模块名

In [69]:
# 调用部分模块语句：From…import 语句

from testmodel2 import f2

print(f2([2,3,4,5,5,5,6,6,4,4,4,4],4))
#print(f3(10,2,10))
# 单独导入模块的部分功能，但无法使用其他未导入模块功能

In [70]:
# python标准模块 —— random随机数

import random

x = random.random()
y = random.random()
print(x,y*10)
# random.random()随机生成一个[0:1)的随机数

m = random.randint(0,10)
print(m)
# random.randint()随机生成一个[0:10]的整数

st1 = random.choice(list(range(10)))
st2 = random.choice('abcdnehgjla')
print(st1,st2)
# random.choice()随机获取()中的一个元素，()种必须是一个有序类型

lst = list(range(20))
sli = random.sample(lst,5)
print(sli)
# random.sample(a,b)随机获取a中指定b长度的片段，不会改变原序列

lst = [1,3,5,7,9,11,13]
random.shuffle(lst)
print(lst)
# random.shuffle(list)将一个列表内的元素打乱

In [73]:
# python标准模块 —— time时间模块

import time

for i in range(2):
    print('hello')
    time.sleep(1)
# time.sleep()程序休息()秒

print(time.ctime())
print(type(time.ctime()))
# 将当前时间转换为一个字符串

print(time.localtime())
print(type(time.localtime()))
# 将当前时间转为当前时区的struct_time
# wday 0-6表示周日到周六
# ydat 1-366 一年中的第几天
# isdst 是否为夏令时，默认为-1

print(time.strftime('%Y-%m-%d %H:%M:%S',time.localtime()))
# time.strftime(a,b)
# a为格式化字符串格式
# b为时间戳，一般用localtime()

%y 两位数的年份表示（00-99）

%Y 四位数的年份表示（000-9999）

%m 月份（01-12）

%d 月内中的一天（0-31）

%H 24小时制小时数（0-23）

%I 12小时制小时数（01-12）

%M 分钟数（00=59）

%S 秒（00-59）

%a 本地简化星期名称

%A 本地完整星期名称

%b 本地简化的月份名称

%B 本地完整的月份名称

%c 本地相应的日期表示和时间表示

%j 年内的一天（001-366）

%p 本地A.M.或P.M.的等价符

%U 一年中的星期数（00-53）星期天为星期的开始

%w 星期（0-6），星期天为星期的开始

%W 一年中的星期数（00-53）星期一为星期的开始

%x 本地相应的日期表示

%X 本地相应的时间表示

%Z 当前时区的名称

%% %号本身

# 7. 数据读写
## 7.1 文件对象声明及基本操作

* 另一种数据格式：文件/文档

In [71]:
# 本地文件的界定：指向一个本地存储的文件，是一个链接或者一个映射

path1 = 'C:/Users/Hjx/Desktop/text.txt'  # 单个反斜杠：/
path2 = 'C:\\Users\\Hjx\\Desktop\\text.txt'  # 两个斜杠：\\（第一个\是转义符）
path3 = r'C:\Users\Hjx\Desktop\text.txt'  # r用于防止字符转义
# 路径书写格式
print(path1)
print(path2)
print(path3)

In [72]:
# 读取文件：open语句

f = open(path2, 'r')
print(type(f))
print(f)
print(f.read())
print('读取完毕')
# open('路径', '模式', enconding = '编码' )
# 模式：r：读取文件，默认；w：写入；rw：读取+写入；a：追加
# 简答的读取方法：.read() → 读取后，光标将会留在读取末尾

print(f.read()) 
print('读取为空')
# 运行第一次.read()之后，光标位于末尾，再次读取输出为空

f.seek(0)
print(f.read()) 
print('第二次读取')
# 所以现在用 f.seek(0) 来移动光标

f.close()
# print(f.read())   # 关闭后无法读取
# 关闭文件链接  f.close()，养成一个好习惯

## 7.2 系统模块下的路径操作

* os 模块：提供了非常丰富的方法用来处理文件和目录

In [74]:
# os模块：系统模块 - 常用命令

import os  # 导入模块

print(os.name)  # 输出字符串指示正在使用的平台。如果是window 则用'nt'表示，对Linux/Unix用户，它是'posix'。

print(os.getcwd())  # 函数得到当前工作目录，即当前Python脚本工作的目录路径。

print(os.listdir())  # 返回指定目录下的所有文件和目录名。

#os.chdir('C:\\Users\\Hjx\\Desktop\\' )  # 切换到目标路径
#print(os.getcwd()) 

#os.remove('text2.txt')  # 删除一个文件

print(os.path.split('C:\\Users\\Hjx\\Desktop\\text.txt'))  # 函数返回一个路径的目录名和文件名

print(os.path.exists('C:\\Users\\Hjx\\Desktop\\heheh.txt'))
print(os.path.exists('C:\\Users\\Hjx\\Desktop\\'))
# 用来检验给出的路径是否真地存在，这里不存在改文件故False，但路径存在故True

In [75]:
# 相对路径和绝对路径

os.chdir('C:\\Users\\Hjx\\Desktop\\' )
f2 = open('text.txt','r')
print(f2.read())
# 在申明了目录之后，就可以直接输出文件名了 → 相对路径

## 7.3 文件的读取与写入

* os 模块：提供了非常丰富的方法用来处理文件和目录

In [77]:
# 文件读取 - read

f = open('C:\\Users\\Hjx\\Desktop\\text.txt', 'r')
print(f.read())
# 直接读取（这里光标默认会在最末尾）

In [78]:
f.seek(0)
print(f.read(2))
# f.read(n):n代表读取多少个字符

In [79]:
f.seek(0)
print(f.readline())
print(f.readline())
print(f.readline(4))
# 读取行到字符串，一次性读取一行
# f.readline(n)：读取该行的前n个字符

In [80]:
f.seek(0)
for line in f.readlines():
    print(type(line),line)
# 遍历一个文件：for语句+f.readlines()

小作业

利用百度POI小插件爬取一些poi数据，然后存成txt，再用python读取，编写成一个json形式（列表字典）：

[{'name':'...', 'lng':..., 'lat':..., 'address':'...'},{...},...,]

In [ ]:
# 答案代码

path = 'C:\\Users\\Hjx\\Desktop\\jiuba.txt'  # 创建路径变量
f = open(path,'r')  # 读取txt文件
m = []  # 新建一个空列表，用于存储数据
n = 0

for line in f.readlines():
    n += 1
    st1 = line.split(':')
    name = st1[0]   # 先以:拆分，筛选出name文本
    information = st1[1]  # 第二部分包括lng，lat，address
    st2 = information.split(',')
    lng = float(st2[0])  # 筛选出lng
    lat = float(st2[1])  # 筛选出lat
    ad = st2[2].strip()  # 筛选出address，去除末尾空格
    data = [['name',name],['lng',lng],['lat',lat],['address',ad]]  # 做成嵌套列表
    m.append(dict(data))  # 生成字典，并追加如列表m
    #print('成功转换%i个数据'%n)

print(m)
print('\n数据转换完成！总共转换%i个数据'%n)

In [ ]:
# 文件写入 - write

path =  'C:\\Users\\Hjx\\Desktop\\test_write.txt' 
f = open(path, 'w', encoding = 'utf8')
f.write('hello world!')
# 执行后直接创建“test_write.txt”文件，并写入“hello world！”
# 12代表输入了12个字符，但是文件内真的有写入吗？打开看看

In [ ]:
f.close()
# 为什么没有？因为刚才是在内存里运行，我们需要再加上close才能写入

In [ ]:
path =  'C:\\Users\\Hjx\\Desktop\\' 
f = open(path + 'test_write2.txt', 'w', encoding = 'utf8')
lst = ['a','b','c','d','e']
f.writelines(lst)
f.close()
# f.writelines(list)：依次把列表元素写入，但这里有没有提行？
# 如果想每一个元素提行怎么办？ → 每个元素后都有"\n"

In [ ]:
path =  'C:\\Users\\Hjx\\Desktop\\' 
f = open(path + 'test_write2.txt', 'w', encoding = 'utf8')
lst = ['a','b','c','d','e']

for i in range(len(lst)):
    lst[i] = lst[i] + '\n'
    
f.writelines(lst)
f.close()
# 通过遍历给每个元素最后加上"\n"

小作业

两个列表[1~10],[a~j]，写入一个txt，变成以下格式

1,a

2,b

3,c

...

In [ ]:
# 答案代码

n = list(range(1,11))
v = ['a','b','c','d','e','f','g','h','i','j']

f = open( 'C:\\Users\\Hjx\\Desktop\\test_write3.txt', 'w', encoding = 'utf8')
m = []

for i in range(len(n)):
    f.writelines([str(n[i]),',',v[i] + '\n'])

f.close()
print('finished!')

## 7.4 pickle模块的运用

* pickle提供了一个简单的持久化功能。可以将对象以文件的形式存放在磁盘上。

* python的pickle模块实现了基本的数据序列和反序列化

* 通过pickle模块的序列化操作我们能够将程序中运行的对象信息保存到文件中去，永久存储

* 通过pickle模块的反序列化操作，我们能够从文件中创建上一次程序保存的对象。

* pickle.dump() / pickle.load()

In [ ]:
# 存储：pickle.dump(obj, file, [,protocol])

import pickle

data = {'a':[1,2,3,4], 'b':('string','abc'), 'c':'hello'}
print(data)
# 创建一个字典变量data

pic = open( 'C:\\Users\\Hjx\\Desktop\\data.pkl', 'wb')
# 以二进制来存储：rb, wb, wrb, ab

pickle.dump(data,pic)
pic.close()
# 将一个字典数据存成了pkl文件

In [ ]:
# 读取：pickle.load(file)

f = open( 'C:\\Users\\Hjx\\Desktop\\data.pkl', 'rb')

st = pickle.load(f)
print(st)
# 直接读取pkl文件内的数据，该数据为一个字典